File: icpc/2018_RPC01/G.py
import random

class Pos:
    def __init__(self, x, y, z):
        self.f = [x, y, z]

    def __str__(self):
        return '(' + ' '.join(str(x) for x in self.f) + ')'

    def __eq__(self, other):
        return self.f == other.f

    def __setitem__(self, i, val):
        self.f[i] = val

    def __getitem__(self, i):
        return self.f[i]

    def randomMove(self, foe):
        i = random.choice([0,1,2])
        v = random.choice([-1,1])
        self[i] += v
        if self == foe:
            self[i] -= v
            self.randomMove(foe)

    def move(self, target, foe):
        if self == target:
            return
        positions = list(range(3))
        random.shuffle(positions)
        for i in positions:
            if not self[i] == target[i]:
                d = target[i] - self[i]
                self[i] += (d > 0) - (d < 0)
                if (self == foe):
                    print('collision: ', self, foe)
                    self[i] -= (d > 0) - (d < 0)
                    self.randomMove(foe)
                else:
                    break


in1 = list(map(int, input().split(' ')))
in2 = list(map(int, input().split(' ')))

a, at = Pos(*in1[:3]), Pos(*in1[3:])
b, bt = Pos(*in2[:3]), Pos(*in2[3:])

print(a, b)
#for i in range(2000):
    #a.randomMove(b)
    #b.randomMove(a)
    #print(a, b)
while not (a == at and b == bt):
    a.move(at, b)
    b.move(bt, a)
    print(a, b)

File: icpc/2018_RPC01/G.cpp
#include <bits/stdc++.h>
using namespace std;

struct Pos {
    int x, y, z;
    bool operator == (const Pos& other) {
        return (other.x == x && other.y == y && other.z == z);
    }
    void move(const Pos& other, const Pos& foe) {
        if (*this == other) return;

        int f[3] {x, y, z};
        int g[3] {other.x, other.y, other.z};
        vector<int> positions = {0,1,2};
        random_shuffle(begin(positions), end(positions));

        for (auto i : positions) {
            if (f[i] != g[i]) {
                int d = (g[i] - f[i]);
                f[i] += ((d > 0) - (d < 0));
                if (f[0] == foe.x && f[1] == foe.y && f[2] == foe.z) {
                    f[i] -= ((d > 0) - (d < 0));
                } else {
                    break;
                }
            }
        }
        tie(x, y, z) = {f[0], f[1], f[2]};
    }
} a, b, at, bt;

ostream& operator <<(ostream& os, const Pos& p) {
    return os << "(" << p.x << ", " << p.y << ", " << p.z << ")";
}

int main(void) {
    cin >> a.x >> a.y >> a.z >> at.x >> at.y >> at.z;
    cin >> b.x >> b.y >> b.z >> bt.x >> bt.y >> bt.z;

    cout << a << ' ' << b << '\n';
    while (!(a == at && b == bt)) {
        a.move(at, b);
        b.move(bt, a);
        cout << a << ' ' << b << '\n';
    }
    return 0;
}
